### PR TITLE
Chapter 12: Password reset

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/assets/stylesheets/password_resets.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/password_resets.scss]: remove file that never used

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user,only: :edit, :update
+  before_action :valid_user, only: :edit, :update
+  before_action :check_expiration, only: :edit, :update
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  before_action :get_user, :valid_user, :check_expiration, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  def create
+    @user = User.find_by email: params[:password_reset][:email].downcase
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = "Email sent with password reset instructions"
+      redirect_to root_url
+    else
+      flash.now[:danger] = "Email address not found"
+      render "new"
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      render :new
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        redirect_to root_path
+      end
+    end
+
+    def check_expiration
+      return unless @user.password_reset_expired?
+        flash[:danger] = "Password reset has expired."
+        redirect_to new_password_reset_path
+      end
+    end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove 2 spaces

File: app/helpers/password_resets_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/password_resets_helper.rb]: remove file that never used

File: app/mailers/user_mailer.rb [hunk 6]
```diff
@@ -4,9 +4,8 @@ def account_activation(user)
     mail to: user.email, subject: "Account activation"
   end
 
-  def password_reset
-    @greeting = "Hi"
-
-    mail to: "to@example.org"
+  def password_reset(user)
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove ()

File: app/models/user.rb [hunk 7]
```diff
@@ -1,7 +1,11 @@
 class User < ApplicationRecord
   USERS_PARAMS = %i(name email password password_confirmation).freeze
 
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline app/models/user.rb]: resolve conflict

File: app/views/password_resets/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.submit "Submit", class: "btn btn-primary" %>
+    <% end %>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb]: add an empty line 13

**Comment by baaohuii** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
+  end
+
+    def valid_user
+    return if @user && @user.activated? && @user.authenticated? (:reset, params[:id])
+      redirect_to root_path
+    end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  def valid_user
    return if @user && @user.activated? && @user.authenticated? (:reset, params[:id])
  
    flash[:danger] = "Some messages"
    redirect_to root_path
```

File: app/controllers/password_resets_controller.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def check_expiration
+    return unless @user.password_reset_expired?
+      flash[:danger] = "Password reset has expired."
+      redirect_to new_password_reset_path
+    end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    return unless @user.password_reset_expired?
    
    flash[:danger] = "Password reset has expired."
    redirect_to new_password_reset_path
```

File: app/models/user.rb [hunk 11]
```diff
@@ -61,10 +51,34 @@ def forget
...
+    self.reset_token = User.new_token
+    update_attributes reset_digest: User.digest(reset_token), reset_sent_at: Time.zone.now
+  end
+
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (saobv-2223) [inline app/models/user.rb]: move 2 to settings.yml file and call it here

File: config/routes.rb [hunk 12]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+  get 'password_resets/new'
+  get 'password_resets/edit'
```
Comment (saobv-2223) [inline config/routes.rb]: using "

**Comment by baaohuii** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      flash.now[:danger] = "Email address not found"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].blank?
+      flash[:danger] = "Password cannot be empty"
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      flash.now[:danger] = "Password cannot be empty"
```

File: app/views/password_resets/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,14 @@
+<% provide :title, "Reset password" %>
+<h1>Reset password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: password_reset_path params[:id]) do |f| %>
```
Comment (chungpv-1008) [inline app/views/password_resets/edit.html.erb]:
```suggestion
    <%= form_for @user, url: password_reset_path(params[:id]) do |f| %>
```

File: app/views/password_resets/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, "Forgot password") %>
+<h1>Forgot password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
```
Comment (chungpv-1008) [inline app/views/password_resets/new.html.erb]:
```suggestion
    <%= form_for :password_reset, url: password_resets_path do |f| %>
```

File: app/views/user_mailer/password_reset.text.erb [hunk 16]
```diff
@@ -1,3 +1,8 @@
-User#password_reset
+To reset your password click the link below:
 
-<%= @greeting %>, find me in app/views/user_mailer/password_reset.text.erb
+<%= edit_password_reset_url(@user.reset_token, email: @user.email) %>
```
Comment (chungpv-1008) [inline app/views/user_mailer/password_reset.text.erb]:
```suggestion
<%= edit_password_reset_url @user.reset_token, email: @user.email %>
```

**Comment by baaohuii** [comment]
ready

File: config/routes.rb [hunk 17]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+  get "password_resets/new"
+  get "password_resets/edit"
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
```

File: config/routes.rb [hunk 18]
```diff
@@ -10,4 +12,5 @@
   get "users/show"
   resources :users
   resources :account_activations, only: :edit
+  resources :password_resets, except: %i(index show destroy)
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
  resources :password_resets
```

**Review by chungpv-1008 — COMMENTED** [review]
Please fix comment, then merge this pull request yourself

**Pull request merged**